feat: adjust plugins to match other elements

higor-melo opened the pull request:
This will also allow the usage of the flag
RockGazebo::Syskit.scope_device_name_with_links_and_submodels to fix the plugins being defined at every level of the model hierarchy instead of only at the last one

higor-melo committed:
feat: adjust plugins to match other elements

This will also allow the usage of the flag
RockGazebo::Syskit.scope_device_name_with_links_and_submodels
to fix the plugins being defined at every level of the
model hierarchy instead of only at the last one

doudou commented on lib/sdf/model.rb:
```diff
@@ -77,7 +85,6 @@ def initialize(xml = REXML::Element.new("model"), parent = nil)
             joints.each do |name, joint_xml|
                 @joints[name] = Joint.new(joint_xml, self)
             end
```
Given that you're cleaning that up ... please take care of the joints as well.

higor-melo replied:
I created the tests and them adapted it, however the i don't like how the code is written. I feel like i'm it could be better, can you double check?

higor-melo committed:
feat: implement direct joints

jhonasiv commented on test/test_model.rb:
```diff
@@ -300,6 +298,43 @@ def models_dir
+            assert root.enum_for(:each_direct_joint).to_a.empty?
+        end
+
+        it "yields the joints no indirect joints" do
```
doesnt yield the submodel's joints

higor-melo committed:
chore: fix test name

doudou commented on lib/sdf/model.rb:
```diff
@@ -77,7 +89,9 @@ def initialize(xml = REXML::Element.new("model"), parent = nil)
             joints.each do |name, joint_xml|
                 @joints[name] = Joint.new(joint_xml, self)
             end
-            @plugins = plugins.map { |child| Plugin.new(child, self) }
+            direct_joints.each do |name, joint_xml|
+                @direct_joints[name] = Joint.new(joint_xml, self)
+            end
```
You have moved this code for Plugin to the main loop, and Link is already doing it there as well ... why keep joint here ?

higor-melo replied:
`joints` needs the link to find the parent

higor-melo committed:
chore: improve code clarity

Use direct links to create global variables
Use transform_values and return value to the global variable
Use direct joints to create the @joints
Create direct_frames as well

higor-melo committed:
feat: add each direct frame tests and method

higor-melo committed:
feat: add direct models

This will be used to avoid setting up the same
orogen model multiple times.
Unit tests pending

higor-melo committed:
fix: maintain old `each_model_with_name` behavior

By merging the values afterward the test:
`enumerates its children model's models`
would break. Upon investigation I was able
to assert that the `.merge!` would in fact
keep the values of the other hash instead
of joining them, causing it to break.

doudou reviewed (commented):
As discussed, process the submodels in the same way than the other items

higor-melo committed:
chore: adapt models to match other elements

higor-melo committed:
fix: rubocop
